ocaml/xapi-guard/lib/lwt_bounded_stream.ml

```diff
@@ -0,0 +1,48 @@
+(*
+ * Copyright (c) 2012 Citrix Systems
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+let ( let* ) = Lwt.bind
+
+type 'a t = {stream: 'a Lwt_stream.t; capacity: int; size: int ref}
+
+let create capacity =
+  let stream, stream_push = Lwt_stream.create () in
+  let t = {stream; capacity; size= ref 0} in
+  let push = function
+    | Some _ when !(t.size) > t.capacity ->
+        None
+    | None ->
+        stream_push None ; Some ()
+    | elem ->
+        stream_push elem ; incr t.size ; Some ()
```

Comment:
What if pushing results in switching to the other lwt promise that immediately consumes the value (thus potentially making the `size` temporarily negative?). Neither ordering between and incr is correct here without introducing explicit locking around pushing and reading available elements.

Reply:
(This could probably be tested with quickcheck to check that size never goes negative and that t.size actually matches the number of elements in the stream, it'd be better if we could somehow query the stream for how many elements it has...)

Reply:
In the disk cache module, queue usage is surrounded by a mutex to ensure calls to the queue module are serialized and size can't be negative. You have a point that the data structure should integrate this.